webapp/src/DOMJudgeBundle/Service/ScoreboardService.php

```diff
@@ -357,6 +358,35 @@ public function calculateScoreRow(Contest $contest, Team $team, Problem $problem
             }
         }
 
+        // See if this submission was the first to solve this problem
+        // (only relevant if it was correct in the first place)
+        $solvedFirst = false;
+        if ($correctJury) {
+            $params = [
+                ':cid' => $contest->getCid(),
+                ':probid' => $problem->getProbid(),
+                ':teamSortOrder' => $team->getCategory()->getSortorder(),
+                ':submitTime' => $absSubmitTime,
+                ':correctResult' => Judging::RESULT_CORRECT,
+            ];
+
+            // Find out how many valid submissions were submitted earlier
+            // that have a valid judging that is correct, or are awaiting judgement.
+            // Only if there are 0 found, we are definitely the first to solve this problem.
+            $solvedFirst = 0 == $this->entityManager->getConnection()->fetchColumn('
+                SELECT count(*) FROM submission s
+                    LEFT JOIN judging j USING (submitid)
+                    LEFT JOIN team t USING(teamid)
+                    LEFT JOIN team_category tc USING (categoryid)
+                WHERE s.valid = 1 AND
+                    ((j.valid = 1 AND ( j.rejudgingid IS NULL AND (j.result IS NULL OR j.result = :correctResult))) OR
```

Comment:
Wow, this is quite complex. I wonder if it makes sense to split this up even further and add comments for:
- in progress judging
- correct result
- queued judging
or similar.

Reply:
Yes, there’s quite some things to account for. Will see if I can make it more transparent.

Reply:
I haven't read the code in full, but is this still keeping track of the first pending or correct submission? Then we'd not need this complex query, but just check whether the currrent submission is the first pending/correct one, and if so, it's the FTS, right?

Reply:
We don't keep any other administration of whether this is the first one for any problem. Because the scorecache is only updated one cell (problem,team) at a time, any solution would always need to do some kind of query on all other team's submissions for this problem.

Indeed the previous pr #509 kept some track of it, namely it changed the definition of the 'solved'-time cell, then stored this information in the cache, and then on scoreboard display check these values to see if it's lowest. This approach does not rely on any other part doing further calculations.